In [1]:
# Import Dependancies
import pandas as pd
from pathlib import Path

### Step 1: filter down the data
We want at least as much data before the pandemic when compared to the pnademic's duration

In [2]:
# Create a file path
path = 'Resources/Crimes_-_2001_to_Present.csv'

# Read csv file with pandas
df = pd.read_csv(path, encoding='ISO-8859-1')

In [5]:
# Filter down data to 2016 and above
df = df[df['Year'] > 2016]

In [6]:
# Check the dataframe
df.head()

ID Case Number                    Date                      Block  \
1     25953    JE240540  05/24/2021 03:06:00 PM        020XX N LARAMIE AVE   
2     26038    JE279849  06/26/2021 09:24:00 AM      062XX N MC CORMICK RD   
3  13279676    JG507211  11/09/2023 07:30:00 AM           019XX W BYRON ST   
4  13274752    JG501049  11/12/2023 07:59:00 AM  086XX S COTTAGE GROVE AVE   
6  13203321    JG415333  09/06/2023 05:00:00 PM           002XX N Wells st   

   IUCR     Primary Type                                        Description  \
1  0110         HOMICIDE                                FIRST DEGREE MURDER   
2  0110         HOMICIDE                                FIRST DEGREE MURDER   
3  0620         BURGLARY                                     UNLAWFUL ENTRY   
4  0454          BATTERY  AGGRAVATED P.O. - HANDS, FISTS, FEET, NO / MIN...   
6  1320  CRIMINAL DAMAGE                                         TO VEHICLE   

                     Location Description  Arrest  Domestic  ...  Ward  \
1                                  STREET    True     False  ...  36.0   
2                             PARKING LOT    True     False  ...  50.0   
3                               APARTMENT   False     False  ...  47.0   
4                      SMALL RETAIL STORE    True     False  ...   6.0   
6  PARKING LOT / GARAGE (NON RESIDENTIAL)   False     False  ...  42.0   

   Community Area  FBI Code  X Coordinate Y Coordinate  Year  \
1            19.0       01A     1141387.0    1913179.0  2021   
2            13.0       01A     1152781.0    1941458.0  2021   
3             5.0        05     1162518.0    1925906.0  2023   
4            44.0       08B     1183071.0    1847869.0  2023   
6            32.0        14     1174694.0    1901831.0  2023   

               Updated On   Latitude  Longitude                       Location  
1  11/18/2023 03:39:49 PM  41.917838 -87.755969  (41.917838056, -87.755968972)  
2  11/18/2023 03:39:49 PM  41.995219 -87.713355  (41.995219444, -87.713354912)  
3  11/18/2023 03:39:49 PM  41.952345 -87.677975  (41.952345086, -87.677975059)  
4  12/09/2023 03:41:24 PM  41.737751 -87.604856  (41.737750767, -87.604855911)  
6  11/04/2023 03:40:18 PM  41.886018 -87.633938  (41.886018055, -87.633937881)  

[5 rows x 22 columns]

In [ ]:
# Output the filtered dataframe to a csv file
df.to_csv('Resources/Crimes_-_2016_to_present.csv', index=False)

### Step 2: Clean the dataset
Our goal for this step was to have the cleanest dataset possible with no missing values. This process involved checking the data structure, getting rid of null values, whether that is through removing the entire row or replacing that value with an alternative value that would aid in the analysis process. 

In [ ]:
# Create file path to import filtered csv
path = 'Resources/Crimes_-_2016_to_present.csv'

# Read csv file with pandas
df = pd.read_csv(path, encoding='ISO-8859-1')

In [ ]:
df.head()

ID Case Number                    Date                      Block  \
0     25953    JE240540  05/24/2021 03:06:00 PM        020XX N LARAMIE AVE   
1     26038    JE279849  06/26/2021 09:24:00 AM      062XX N MC CORMICK RD   
2  13279676    JG507211  11/09/2023 07:30:00 AM           019XX W BYRON ST   
3  13274752    JG501049  11/12/2023 07:59:00 AM  086XX S COTTAGE GROVE AVE   
4  13203321    JG415333  09/06/2023 05:00:00 PM           002XX N Wells st   

   IUCR     Primary Type                                        Description  \
0  0110         HOMICIDE                                FIRST DEGREE MURDER   
1  0110         HOMICIDE                                FIRST DEGREE MURDER   
2  0620         BURGLARY                                     UNLAWFUL ENTRY   
3  0454          BATTERY  AGGRAVATED P.O. - HANDS, FISTS, FEET, NO / MIN...   
4  1320  CRIMINAL DAMAGE                                         TO VEHICLE   

                     Location Description  Arrest  Domestic  ...  Ward  \
0                                  STREET    True     False  ...  36.0   
1                             PARKING LOT    True     False  ...  50.0   
2                               APARTMENT   False     False  ...  47.0   
3                      SMALL RETAIL STORE    True     False  ...   6.0   
4  PARKING LOT / GARAGE (NON RESIDENTIAL)   False     False  ...  42.0   

   Community Area  FBI Code  X Coordinate Y Coordinate  Year  \
0            19.0       01A     1141387.0    1913179.0  2021   
1            13.0       01A     1152781.0    1941458.0  2021   
2             5.0        05     1162518.0    1925906.0  2023   
3            44.0       08B     1183071.0    1847869.0  2023   
4            32.0        14     1174694.0    1901831.0  2023   

               Updated On   Latitude  Longitude                       Location  
0  11/18/2023 03:39:49 PM  41.917838 -87.755969  (41.917838056, -87.755968972)  
1  11/18/2023 03:39:49 PM  41.995219 -87.713355  (41.995219444, -87.713354912)  
2  11/18/2023 03:39:49 PM  41.952345 -87.677975  (41.952345086, -87.677975059)  
3  12/09/2023 03:41:24 PM  41.737751 -87.604856  (41.737750767, -87.604855911)  
4  11/04/2023 03:40:18 PM  41.886018 -87.633938  (41.886018055, -87.633937881)  

[5 rows x 22 columns]

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2202529 entries, 0 to 2202528
Data columns (total 22 columns):
 #   Column                Dtype  
---  ------                -----  
 0   ID                    int64  
 1   Case Number           object 
 2   Date                  object 
 3   Block                 object 
 4   IUCR                  object 
 5   Primary Type          object 
 6   Description           object 
 7   Location Description  object 
 8   Arrest                bool   
 9   Domestic              bool   
 10  Beat                  int64  
 11  District              float64
 12  Ward                  float64
 13  Community Area        float64
 14  FBI Code              object 
 15  X Coordinate          float64
 16  Y Coordinate          float64
 17  Year                  int64  
 18  Updated On            object 
 19  Latitude              float64
 20  Longitude             float64
 21  Location              object 
dtypes: bool(2), float64(7), int64(3), object(1

In [ ]:
# Check for row and columns with null values
crime_dfrows_with_nan = df[df.isnull().any(axis=1)]
crime_dfrows_with_nan.head()

ID Case Number                    Date                   Block  \
11   13579934    JH406506  08/20/2024 08:00:00 AM      043XX S ALBANY AVE   
35   12416972    JE293535  10/01/2020 12:01:00 AM         031XX W 53RD PL   
41   12978205    JG146937  08/01/2022 10:00:00 PM      028XX S KEELER AVE   
43   13041906    JG203447  10/15/2022 01:30:00 AM  051XX W WRIGHTWOOD AVE   
129  13582742    JH410857  09/01/2023 12:01:00 AM  059XX N RAVENSWOOD AVE   

     IUCR                Primary Type  \
11   1120          DECEPTIVE PRACTICE   
35   1753  OFFENSE INVOLVING CHILDREN   
41   1562                 SEX OFFENSE   
43   0486                     BATTERY   
129  2825               OTHER OFFENSE   

                                  Description Location Description  Arrest  \
11                                    FORGERY                  NaN   False   
35   SEXUAL ASSAULT OF CHILD BY FAMILY MEMBER            RESIDENCE   False   
41           AGGRAVATED CRIMINAL SEXUAL ABUSE            APARTMENT   False   
43                    DOMESTIC BATTERY SIMPLE            APARTMENT    True   
129                   HARASSMENT BY TELEPHONE            APARTMENT   False   

     Domestic  ...  Ward  Community Area  FBI Code  X Coordinate Y Coordinate  \
11      False  ...  12.0            58.0        10     1156421.0    1875687.0   
35       True  ...  14.0            63.0        02           NaN          NaN   
41       True  ...  22.0            30.0        17           NaN          NaN   
43       True  ...  31.0            19.0       08B           NaN          NaN   
129      True  ...  40.0            77.0        26           NaN          NaN   

     Year              Updated On   Latitude  Longitude  \
11   2024  09/02/2024 03:40:51 PM  41.814665 -87.701746   
35   2020  09/14/2023 03:41:59 PM        NaN        NaN   
41   2022  09/14/2023 03:41:59 PM        NaN        NaN   
43   2022  09/14/2023 03:41:59 PM        NaN        NaN   
129  2023  09/02/2024 03:41:30 PM        NaN        NaN   

                          Location  
11   (41.814665041, -87.701746474)  
35                             NaN  
41                             NaN  
43                             NaN  
129                            NaN  

[5 rows x 22 columns]

In [11]:
# Count missing values in each column and display the result
missing_values_count = df.isnull().sum()
print(missing_values_count)


ID                          0
Case Number                 0
Date                        0
Block                       0
IUCR                        0
Primary Type                0
Description                 0
Location Description    11277
Arrest                      0
Domestic                    0
Beat                        0
District                    1
Ward                       53
Community Area              2
FBI Code                    0
X Coordinate            31189
Y Coordinate            31189
Year                        0
Updated On                  0
Latitude                31189
Longitude               31189
Location                31189
dtype: int64


Because the 'X Coordinate', 'Y Coordinate', 'Latitude', 'Longitude', and 'Location' columns all have a missing value count of 31189, I assumed that they are all on the same rows. We will check this.

In [14]:
# Get the total number of rows in the filtered dataframe
total_rows = len(df)
print(total_rows)


2202529


In [15]:
# Calculate the percentage of missing values for the column 'Longitude'
missing_percentage = (df['Longitude'].isnull().sum() / len(df)) * 100
print(missing_percentage)


1.4160539997430226


In [ ]:
# Checking the district count of unique values
district_count = len(df['District'].unique())
district_count

24

In [17]:
# Checking the district count of unique values
iucr_count = len(df['IUCR'].unique())
iucr_count

382

In [18]:
# Drop rows where 'Location' has a missing value
# Because the Location column is just a concatination of both Latitude and Longitude values, it is the best to get rid of first 
# due to the fact that its just repeating values
dropped_rows = df.dropna(subset=['Location'])


In [19]:
# Checking to see how that affected the dataset's null values
dropped_rows_missing_count = dropped_rows.isnull().sum()
print(dropped_rows_missing_count)

ID                         0
Case Number                0
Date                       0
Block                      0
IUCR                       0
Primary Type               0
Description                0
Location Description    7867
Arrest                     0
Domestic                   0
Beat                       0
District                   1
Ward                      51
Community Area             1
FBI Code                   0
X Coordinate               0
Y Coordinate               0
Year                       0
Updated On                 0
Latitude                   0
Longitude                  0
Location                   0
dtype: int64


In [20]:
# Getting rid of Ward (another way to describe the police jurisdiction), Community Area and District null vlaues (there 
# is only on missing vlaue for these columns)
more_dropped = dropped_rows.dropna(subset=['Ward','Community Area','District'])

In [22]:
# Checking to see how this affected the dataset's null values
more_dropped_rows_missing_count = more_dropped.isnull().sum()
print(more_dropped_rows_missing_count)


ID                         0
Case Number                0
Date                       0
Block                      0
IUCR                       0
Primary Type               0
Description                0
Location Description    7867
Arrest                     0
Domestic                   0
Beat                       0
District                   0
Ward                       0
Community Area             0
FBI Code                   0
X Coordinate               0
Y Coordinate               0
Year                       0
Updated On                 0
Latitude                   0
Longitude                  0
Location                   0
dtype: int64


Because 'Location Description is a additional describer of location, we thought it would be fine to fill in those missing values with the word "Unknown"

In [23]:
# Replace NaN values in 'Location Description' and save in a new DataFrame
cleaned_crime_data = more_dropped.copy()  # Create a copy of the original DataFrame
cleaned_crime_data['Location Description'] = cleaned_crime_data['Location Description'].fillna('Unknown')

# Display the first few rows of the cleaned DataFrame
cleaned_crime_data.head()


ID Case Number                    Date                      Block  \
0     25953    JE240540  05/24/2021 03:06:00 PM        020XX N LARAMIE AVE   
1     26038    JE279849  06/26/2021 09:24:00 AM      062XX N MC CORMICK RD   
2  13279676    JG507211  11/09/2023 07:30:00 AM           019XX W BYRON ST   
3  13274752    JG501049  11/12/2023 07:59:00 AM  086XX S COTTAGE GROVE AVE   
4  13203321    JG415333  09/06/2023 05:00:00 PM           002XX N Wells st   

   IUCR     Primary Type                                        Description  \
0  0110         HOMICIDE                                FIRST DEGREE MURDER   
1  0110         HOMICIDE                                FIRST DEGREE MURDER   
2  0620         BURGLARY                                     UNLAWFUL ENTRY   
3  0454          BATTERY  AGGRAVATED P.O. - HANDS, FISTS, FEET, NO / MIN...   
4  1320  CRIMINAL DAMAGE                                         TO VEHICLE   

                     Location Description  Arrest  Domestic  ...  Ward  \
0                                  STREET    True     False  ...  36.0   
1                             PARKING LOT    True     False  ...  50.0   
2                               APARTMENT   False     False  ...  47.0   
3                      SMALL RETAIL STORE    True     False  ...   6.0   
4  PARKING LOT / GARAGE (NON RESIDENTIAL)   False     False  ...  42.0   

   Community Area  FBI Code  X Coordinate Y Coordinate  Year  \
0            19.0       01A     1141387.0    1913179.0  2021   
1            13.0       01A     1152781.0    1941458.0  2021   
2             5.0        05     1162518.0    1925906.0  2023   
3            44.0       08B     1183071.0    1847869.0  2023   
4            32.0        14     1174694.0    1901831.0  2023   

               Updated On   Latitude  Longitude                       Location  
0  11/18/2023 03:39:49 PM  41.917838 -87.755969  (41.917838056, -87.755968972)  
1  11/18/2023 03:39:49 PM  41.995219 -87.713355  (41.995219444, -87.713354912)  
2  11/18/2023 03:39:49 PM  41.952345 -87.677975  (41.952345086, -87.677975059)  
3  12/09/2023 03:41:24 PM  41.737751 -87.604856  (41.737750767, -87.604855911)  
4  11/04/2023 03:40:18 PM  41.886018 -87.633938  (41.886018055, -87.633937881)  

[5 rows x 22 columns]

In [25]:
# Checking primary type occurence count across the entire dataset
# This helps us find out the top 4 crimes that occured in the filtered and cleaned dataset
cleaned_crime_count = cleaned_crime_data.groupby('Primary Type').size()
cleaned_crime_count.sort_values(ascending=False)

Primary Type
THEFT                                489381
BATTERY                              402701
CRIMINAL DAMAGE                      243963
ASSAULT                              179954
DECEPTIVE PRACTICE                   148930
OTHER OFFENSE                        137443
MOTOR VEHICLE THEFT                  130405
BURGLARY                              85359
ROBBERY                               84639
NARCOTICS                             77808
WEAPONS VIOLATION                     61349
CRIMINAL TRESPASS                     47136
OFFENSE INVOLVING CHILDREN            17490
PUBLIC PEACE VIOLATION                10233
SEX OFFENSE                            9693
CRIMINAL SEXUAL ASSAULT                7994
INTERFERENCE WITH PUBLIC OFFICER       7366
HOMICIDE                               6028
CRIM SEXUAL ASSAULT                    5098
ARSON                                  4150
PROSTITUTION                           4028
STALKING                               2682
LIQUOR LAW VIOLATIO

In [ ]:
# Output the newly cleaned dataframe to a csv file
cleaned_crime_data.to_csv('Resources/Cleaned_Crimes_-_2016_to_present.csv')